File: androidApp/src/main/kotlin/com/chaoticware/discrobble/android/auth/AuthCallbackParser.kt
package com.chaoticware.discrobble.android.auth

import android.net.Uri

class AuthCallbackParser {
    fun parse(uri: Uri): AuthCallbackResult {
        val scheme = uri.scheme?.lowercase()
        require(scheme == "discrobble") {
            "The callback URL must use the discrobble scheme."
        }

        require(uri.host?.lowercase() == "auth") {
            "The callback URL must target discrobble://auth/lastfm or discrobble://auth/discogs."
        }

        val provider = AuthProvider.fromRawValue(uri.lastPathSegment?.lowercase().orEmpty())
            ?: throw IllegalArgumentException(
                "The callback URL must target discrobble://auth/lastfm or discrobble://auth/discogs."
            )

        val encodedFragment = uri.encodedFragment.orEmpty()

        if (encodedFragment.isNotEmpty() && !encodedFragment.contains('=')) {
            return invalidCallbackFailure(
                provider = provider,
                message = "The callback URL fragment was not encoded as expected.",
            )
        }

        // The Worker encodes callback fragments as query-string pairs such as
        // #payload=... or #error_code=...&error_message=..., so re-parse the
        // encoded fragment as a synthetic query URI before reading keys.
        val fragmentUri = Uri.parse("discrobble://fragment?$encodedFragment")
        val errorCode = fragmentUri.getQueryParameter("error_code")?.trim().orEmpty()

        if (errorCode.isNotEmpty()) {
            val errorMessage = fragmentUri.getQueryParameter("error_message")
                ?.trim()
                .takeUnless { it.isNullOrEmpty() }
                ?: "The provider auth flow did not complete successfully."

            return AuthCallbackResult.Failure(
                failure = AuthCallbackFailure(
                    provider = provider,
                    code = errorCode,
                    message = errorMessage,
                ),
            )
        }

        val payload = fragmentUri.getQueryParameter("payload")
            ?.trim()
            .orEmpty()

        if (payload.isEmpty()) {
            return invalidCallbackFailure(
                provider = provider,
                message = "The callback URL is missing the encrypted payload fragment.",
            )
        }

        return AuthCallbackResult.Payload(
            callback = PendingAuthCallback(
                provider = provider,
                encryptedPayload = payload,
                receivedAtEpochMillis = System.currentTimeMillis(),
            ),
        )
    }

    private fun invalidCallbackFailure(
        provider: AuthProvider,
        message: String,
    ): AuthCallbackResult {
        return AuthCallbackResult.Failure(
            failure = AuthCallbackFailure(
                provider = provider,
                code = "invalid_callback",
                message = message,
            ),
        )
    }
}

File: iosApp/Discrobble/Auth/AuthCallbackParser.swift
import Foundation

struct AuthCallbackParser {
    func parse(url: URL) throws -> AuthCallbackResult {
        guard url.scheme?.lowercased() == "discrobble" else {
            throw AuthCallbackParserError.unsupportedScheme
        }

        let provider = try parseProvider(from: url)
        let fragment = parseFragment(from: url)

        if let errorCode = fragment["error_code"], !errorCode.isEmpty {
            let errorMessage = fragment["error_message"]?.trimmingCharacters(in: .whitespacesAndNewlines)
            return .failure(
                AuthCallbackFailure(
                    provider: provider,
                    code: errorCode,
                    message: errorMessage ?? "The provider auth flow did not complete successfully."
                )
            )
        }

        guard let payload = fragment["payload"]?.trimmingCharacters(in: .whitespacesAndNewlines), !payload.isEmpty else {
            throw AuthCallbackParserError.missingPayload
        }

        return .payload(
            PendingAuthCallback(
                provider: provider,
                encryptedPayload: payload,
                receivedAt: Date()
            )
        )
    }

    private func parseProvider(from url: URL) throws -> AuthProvider {
        guard url.host?.lowercased() == "auth" else {
            throw AuthCallbackParserError.unsupportedRoute
        }

        let route = url.path.trimmingCharacters(in: CharacterSet(charactersIn: "/")).lowercased()

        guard let provider = AuthProvider(rawValue: route) else {
            throw AuthCallbackParserError.unsupportedRoute
        }

        return provider
    }

    private func parseFragment(from url: URL) -> [String: String] {
        guard let fragment = url.fragment, !fragment.isEmpty else {
            return [:]
        }

        var fragmentComponents = URLComponents()
        fragmentComponents.scheme = "discrobble"
        fragmentComponents.host = "fragment"
        fragmentComponents.percentEncodedQuery = fragment

        return Dictionary(
            fragmentComponents
            .queryItems?
            .compactMap { item in
                guard let value = item.value else {
                    return nil
                }

                return (item.name, value)
            } ?? [],
            uniquingKeysWith: { _, last in last }
        )
    }
}

enum AuthCallbackParserError: LocalizedError {
    case unsupportedScheme
    case unsupportedRoute
    case missingPayload

    var errorDescription: String? {
        switch self {
        case .unsupportedScheme:
            return "The callback URL must use the discrobble scheme."
        case .unsupportedRoute:
            return "The callback URL must target discrobble://auth/lastfm or discrobble://auth/discogs."
        case .missingPayload:
            return "The callback URL is missing the encrypted payload fragment."
        }
    }
}
